File: pset2/vigenere.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <cs50.h>

// define max key length for array purpose
#define MAX 50

int main(int argc, char** argv) {
  // Verifying the number of command-line arguments
  if (argc != 2) {
    printf("Usage: ./vigenere k\n");
    return 1;
  }

  // initialize variables
  int length;
  char plaintext[MAX], tmp;

  // verifying that argv is only alphabetical
  for (int i = 0; argv[1][i] != '\0'; i++) {
    if (isalpha(argv[1][i]) == 0) {
        printf("Usage: ./vigenere k\n");
        return 1;
    }
  }

  // getting the length of the key for the modulo
  length = strlen(argv[1]);

  // getting the input string
  printf("plaintext: ");
  fgets (plaintext, MAX, stdin);

  // initialize key and saving argv1 as a string
  char key[length];
  strcpy(key, argv[1]);

  // alphabet index to check for uppercase and lowercase
  for (int j = 0; j < length; j++)
    key[j] -= (key[j] <= 'Z' ? 'A' : 'a');

  // cipher algorithm to encrypt plaintext
  printf("ciphertext: ");
  for (int k = 0, count = 0, n = strlen(plaintext); k < n; k++) {
    if (isalpha(plaintext[k]) == 0)
      printf("%c", plaintext[k]);
    else {
      tmp = (plaintext[k] <= 'Z' ? 'A' : 'a');
      plaintext[k] = plaintext[k] - tmp;
      printf("%c", (((plaintext[k] + key[count % length]) % 26) + tmp));
      count++;
    }
  }

  // print new line and render success
  printf("\n");
  return 0;
}

File: pset6/vigenere.py
from __future__ import print_function
import cs50
import sys

# encrypt a text with the vigenere method
def main():
    # ensure proper usage
    if len(sys.argv) != 2 or not str.isalpha(sys.argv[1]):
        print("Usage: python vigenere.py k")
        exit(1)

    # get key and its length for modulo
    key = sys.argv[1]
    length = len(key)

    # prompt user for text
    print("plaintext: ", end="")
    plaintext = cs50.get_string()

    # convert to list
    keyList = list(key)
    textList = list(plaintext)

    # alphabet index to check for uppercase and lowercase
    for i in range(length):
        keyList[i] = chr(ord(keyList[i]) - (ord('A') if ord(keyList[i]) <= ord('Z') else ord('a')))

    # cipher algorithm to encrypt plaintext
    print("ciphertext: ", end="")
    count = 0
    for j in range(len(textList)):
        # do not encrypt nonalphabetical character
        if textList[j].isalpha() == False:
            print("{}".format(textList[j]), end="")
        else:
            tmp = ord('A') if ord(textList[j]) <= ord('Z') else ord('a')
            textList[j] = ord(textList[j]) - tmp
            print("{}".format(chr(((textList[j] + ord(keyList[count % length])) % 26) + tmp)), end="")
            count += 1

    # print newline and render success
    print("")
    return 0

# call main
if __name__ == "__main__":
    main()
